[Datumaro] CLI updates

zhiltsov-max opened the pull request:
Added:
- README, user guide, developer guide
- Refactored CLI - parameter names, command names, added command descriptions
- `project info` command
- `model run` command
- VS Code environment updated for Datumaro
- `image_dir` extractor, which creates a dataset from a directory

zhiltsov-max commented:
@nmanovic, @azhavoro, @bsekachev, please, take a look on docs and CLI.

bsekachev commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -1,3 +1,8 @@
+
+# Copyright (C) 2018 Intel Corporation
```
2018-2020?

bsekachev commented:
Hi, 
Some ideas and questions. Maybe I did wrong somewhere, but nevertheless. 

### Installation 
#### The first way
I had done:
```bash 
python3 -m venv .env  # python 3.6.8
. .env/bin/activate
pip install 'git+https://github.com/opencv/cvat#egg=datumaro&subdirectory=datumaro'
datum --help
```

And have got exception:
<details><summary>Traceback</summary>
    <pre>
Traceback (most recent call last):
  File "/home/bsekache/datumaro/.env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 574, in _build_master
    ws.require(__requires__)
  File "/home/bsekache/datumaro/.env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 892, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/home/bsekache/datumaro/.env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 783, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (setuptools 39.0.1 (/home/bsekache/datumaro/.env/lib/python3.6/site-packages), Requirement.parse('setuptools>=41.0.0'), {'tensorboard'})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/bsekache/datumaro/.env/bin/datum", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/home/bsekache/datumaro/.env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3088, in <module>
    @_call_aside
  File "/home/bsekache/datumaro/.env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3072, in _call_aside
    f(*args, **kwargs)
  File "/home/bsekache/datumaro/.env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3101, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/home/bsekache/datumaro/.env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 576, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/home/bsekache/datumaro/.env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 589, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/home/bsekache/datumaro/.env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 778, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'setuptools>=41.0.0' distribution was not found and is required by tensorboard
    </pre>
   </details>

#### The second way is okay (from CVAT dir)
Just need to add after clone and before ``while read -r p; do pip install $p; done < requirements.txt``:
```bash
 cd datumaro
```
By the way, why not just ``pip install -r requirements.txt``? Why such the exotic command to install pip dependencies?

The command ``datum`` doesn't work for me in this installation way.
Only: ``python datum.py`` or ``./datum.py``.

### Add and remove data
According to the documentation this is right way to add data:
```bash
./datum.py source add -p ~/datumproject/ path ~/Downloads/Video\ Share\ \(1\).xml -f cvat
```
But it doesn't work with ``-f`` argument. So, 
```bash
./datum.py source add -p ~/datumproject/ path ~/Downloads/Video\ Share\ \(1\).xml
```
was okay for me. 

There aren't documentation about how to remove data. Only in command line tool.
Commands to add and remove data doesn't look consistently.

```bash
./datum.py source add -p ~/datumproject/ path ~/cvat_intel/cvat_public/data/9/data/0/ # add data
./datum.py source remove -p ~/datumproject/ --name ~/cvat_intel/cvat_public/data/9/data/0/ # remove data
```

In the first case we use positional argument ``path`` in the second case we use optional argument ``--name``.

It would be better if remove was:
```
./datum.py source remove -p ~/datumproject/ path ~/cvat_intel/cvat_public/data/9/data/0/
```

By the way, I can a lot of times enter this command to remove data and every time I get:
```
INFO: Source '/home/bsekache/cvat_intel/cvat_public/data/9/data/0/' has been removed from the project
```
But when I try to export or extract the dataset, _I get the error_:
```
No such file or directory: '/home/bsekache/cvat_intel/cvat_public/data/9/data/0/config.yaml'
```

Is it expected behavior? 

How can I display all sources?

There is info about custom extractors and importers, but I haven't found examples.

### Filter

``Filter`` uses command ``extract``. May be is it better to choose something one?
Can I get just simple statistics without creating a new project?`` It would be great. 

### Import
```bash
./datum.py project import -i ~/Downloads/task_new\ task-2020_01_22_08_56_14-coco -o ~/datumproject_1/ -f coco
```
I get error here:

<details><summary>Traceback:</summary>
    <pre>
2020-01-22 12:53:30,618 INFO: Importing project from '/home/bsekache/Downloads/task_new task-2020_01_22_08_56_14-coco' as 'coco'
2020-01-22 12:53:30,622 ERROR: 'coco'
Traceback (most recent call last):
  File "./datum.py", line 8, in <module>
    sys.exit(main())
  File "/home/bsekache/cvat_intel/cvat_public/datumaro/datumaro/cli/__main__.py", line 99, in main
    return args.command(args)
  File "/home/bsekache/cvat_intel/cvat_public/datumaro/datumaro/cli/contexts/project/__init__.py", line 161, in import_command
    project = Project.import_from(source, args.format)
  File "/home/bsekache/cvat_intel/cvat_public/datumaro/datumaro/components/project.py", line 715, in import_from
    importer = env.make_importer(dataset_format)
  File "/home/bsekache/cvat_intel/cvat_public/datumaro/datumaro/components/project.py", line 240, in make_importer
    return self.importers.get(name)(*args, **kwargs)
  File "/home/bsekache/cvat_intel/cvat_public/datumaro/datumaro/components/project.py", line 60, in get
    return self.items[key] # returns a class / ctor
KeyError: 'coco'
    </pre>
   </details>

I get this coco dataset from CVAT (export as a dataset).

### Diff

> The current project is considered to be "ground truth".

It can be inconvenient. I need to go to the "ground truth" directory (being in venv) and specify the full path to datum.py. 

```bash
python /home/bsekache/cvat_intel/cvat_public/datumaro/datum.py project diff ../2_task_new\ task-2020_01_22_09_01_25-datumaro_project 
```
If I can get ``datum`` from environment, it would not be a problem.
First time it asked me my CVAT credentials. Why? We probably need to describe a reason in user guide.

nmanovic commented on datumaro/datumaro/cli/commands/add.py:
```diff
@@ -0,0 +1,8 @@
+
+# Copyright (C) 2019 Intel Corporation
```
@zhiltsov-max , please update all copyright comments. It should have 2020 or 2019-2020 years

nmanovic commented:
Installation steps should not include steps to create a virtual environment but you can recommend that. For me it works fine without any problems.

```
(datumaro) nmanovic@Nikitas-MacBook-Pro:~/Workspace/cvat$ datum
usage: datum [-h] [--version] [--loglevel LOGLEVEL]
             {project,source,item,model,create,add,remove,export,stats,explain}
             ...
datum: error: the following arguments are required: command, args
```

2. Probably installation steps should be after introduction

3. Please move development section into a separate document: CONTRIBUTING.md. You can reference the document in README.md. Look at the following examples to improve README.md:
https://github.com/tensorflow/tensorflow
https://github.com/scikit-image/scikit-image
https://github.com/pytorch/pytorch

4. Please provide a link on CVAT format as you did for others: https://github.com/opencv/cvat/blob/develop/cvat/apps/documentation/xml_format.md (or on a specific revision)

5. In User Guide. Is it correct?

```
As a python module: The directory containing Datumaro should be in the PYTHONPATH environment variable.

datum --help
python datumaro/ --help
datum —help
```

6. Probably you want to provide more information about XPath (link on the package with more examples) and more examples in your user guide

7. I used the following command to install: `pip install --force-reinstall ‘git+https://github.com/opencv/cvat.git@zm/dm-cli-and-docs-update#egg=datumaro&subdirectory=datumaro'`

It looks like we will have problem in the future with installation of the latest version if you don’t carefully support versioning. Please think how to handle the issue.

8. After installation I tried to run a command and got:

```
(datumaro) nmanovic@Nikitas-MacBook-Pro:/tmp$ datum project import --format voc --input-path ~/Downloads/VOCtrainval_06-Nov-2007.tar
Traceback (most recent call last):
  File "/Users/nmanovic/.virtualenvs/datumaro/bin/datum", line 5, in <module>
    from datumaro import main
ImportError: cannot import name 'main' from 'datumaro' (/Users/nmanovic/.virtualenvs/datumaro/lib/python3.7/site-packages/datumaro/__init__.py)
```

zhiltsov-max commented:
@nmanovic, 
1. done
2. done
3. the examples look too complicated. Done for an extent.
4. done
5. fixed
6. added few more examples
7. added a note in the user guide
8. fixed

@bsekachev,
installation.the first way. Looks like your environment contains outdated setuptools, try to update the package
installation.the second way. This weird setup line comes from the issue with `Cpython` package, which may be required to build another package `pycocotools`, but it can't be done in a single step. `datum` is expected not to be working in this scenario, at least for now.
data sources.`-f` argument. Fixed and removed default value
data sources.`remove`. Added to the docs
data sources.consistency. Both these commands operate on sources by their names, `path` is the type of the source
data sources.how to list all. added `project info` command
filter.extract != filter. Renamed the section in the docs
filter.statistics. Statistics are not implemented yet, but the trivial ones now available in `project info`.
import.wrong importer name. Fixed
diff.current project. The "current" project can be changed by `-p` parameter.
diff.credentials. It must have written it was needed for remote dataset access for a project, which was exported as Datumaro project from CVAT.

nmanovic commented:
Extracted `VOCtrainval_06-Nov-2007.tar`. Run the command below:
```
(datumaro) nmanovic@Nikitas-MacBook-Pro:/tmp$ datum project import -o /tmp/test5 --format voc --input-path /tmp/VOCdataset/VOCdevkit/VOC2007/
2020-01-23 16:56:22,026 INFO: Importing project from '/tmp/VOCdataset/VOCdevkit/VOC2007/' as 'voc'
2020-01-23 16:56:22,031 INFO: Checking the dataset...
2020-01-23 16:56:24,986 ERROR: type object argument after * must be an iterable, not NoneType
Traceback (most recent call last):
  File "/Users/nmanovic/.virtualenvs/datumaro/bin/datum", line 8, in <module>
    sys.exit(main())
  File "/Users/nmanovic/.virtualenvs/datumaro/lib/python3.7/site-packages/datumaro/cli/__main__.py", line 99, in main
    return args.command(args)
  File "/Users/nmanovic/.virtualenvs/datumaro/lib/python3.7/site-packages/datumaro/cli/contexts/project/__init__.py", line 168, in import_command
    dataset = project.make_dataset()
  File "/Users/nmanovic/.virtualenvs/datumaro/lib/python3.7/site-packages/datumaro/components/project.py", line 728, in make_dataset
    return ProjectDataset(self)
  File "/Users/nmanovic/.virtualenvs/datumaro/lib/python3.7/site-packages/datumaro/components/project.py", line 486, in __init__
    for item in source:
  File "/Users/nmanovic/.virtualenvs/datumaro/lib/python3.7/site-packages/datumaro/components/extractors/voc.py", line 122, in __iter__
    for item in subset:
  File "/Users/nmanovic/.virtualenvs/datumaro/lib/python3.7/site-packages/datumaro/components/extractors/voc.py", line 34, in __iter__
    yield self._parent._get(item_id, self._name)
  File "/Users/nmanovic/.virtualenvs/datumaro/lib/python3.7/site-packages/datumaro/components/extractors/voc.py", line 132, in _get
    annotations = self._get_annotations(item_id)
  File "/Users/nmanovic/.virtualenvs/datumaro/lib/python3.7/site-packages/datumaro/components/extractors/voc.py", line 235, in _get_annotations
    group=obj_id))
TypeError: type object argument after * must be an iterable, not NoneType
```

azhavoro commented:
Project was exported from CVAT
1. datum project export -f voc -- --save-images
```
▶ datum project export -f voc -- --save-images
2020-01-23 16:46:27,544 INFO: Loading the project...
2020-01-23 16:46:27,643 INFO: Exporting the project...
2020-01-23 16:46:27,643 ERROR: Unknown format code 'd' for object of type 'str'
Traceback (most recent call last):
  File "/home/andrey/workspace/intel/cvat_public/test_datum/venv/bin/datum", line 11, in <module>
    load_entry_point('datumaro', 'console_scripts', 'datum')()
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/cli/__main__.py", line 99, in main
    return args.command(args)
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/cli/contexts/project/__init__.py", line 301, in export_command
    **filter_args)
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/project.py", line 672, in export_project
    converter(dataset, save_dir)
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/converters/voc.py", line 524, in __call__
    converter.convert()
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/converters/voc.py", line 60, in convert
    self.save_subsets()
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/converters/voc.py", line 123, in save_subsets
    data = item.image
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/project.py", line 307, in image
    return self._image()
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/project.py", line 420, in <lambda>
    return lambda: item.image
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/project.py", line 309, in image
    return self._item.image
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/extractor.py", line 533, in image
    return self._image()
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/util/image.py", line 146, in __call__
    image = self.loader(self.path)
  File "/home/andrey/workspace/intel/cvat_public/test_datum/.datumaro/extractors/cvat_rest_api_task_images.py", line 39, in <lambda>
    lambda item_id: self._image_loader(item_id, self))
  File "/home/andrey/workspace/intel/cvat_public/test_datum/.datumaro/extractors/cvat_rest_api_task_images.py", line 81, in _image_loader
    if not extractor._is_image_cached(item_id):
  File "/home/andrey/workspace/intel/cvat_public/test_datum/.datumaro/extractors/cvat_rest_api_task_images.py", line 42, in _is_image_cached
    return osp.isfile(self._image_local_path(item_id))
  File "/home/andrey/workspace/intel/cvat_public/test_datum/.datumaro/extractors/cvat_rest_api_task_images.py", line 35, in _image_local_path
    'task_{}_frame_{:06d}.jpg'.format(task_id, item_id))
ValueError: Unknown format code 'd' for object of type 'str'
```

2. datum project export -f voc -e '/item[image/width < image/height]'
```
datum project export -f voc -e '/item[image/width < image/height]'
2020-01-23 16:51:54,869 INFO: Loading the project...
2020-01-23 16:51:54,965 INFO: Exporting the project...
2020-01-23 16:51:54,965 ERROR: XPathDatasetFilter() got an unexpected keyword argument 'remove_empty'
Traceback (most recent call last):
  File "/home/andrey/workspace/intel/cvat_public/test_datum/venv/bin/datum", line 11, in <module>
    load_entry_point('datumaro', 'console_scripts', 'datum')()
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/cli/__main__.py", line 99, in main
    return args.command(args)
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/cli/contexts/project/__init__.py", line 301, in export_command
    **filter_args)
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/project.py", line 670, in export_project
    remove_empty=remove_empty)
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/project.py", line 406, in extract
    return self.transform(XPathDatasetFilter, filter_expr, **kwargs)
  File "/home/andrey/workspace/intel/cvat_public/test_datum/datumaro/datumaro/components/extractor.py", line 616, in transform
    return method(self, *args, **kwargs)
TypeError: XPathDatasetFilter() got an unexpected keyword argument 'remove_empty'
```
3. It would be nice to add more examples for filter expression (especially for more complicated cases, ie label == 'dog' or label == 'cat' if it's possible) in the section https://github.com/opencv/cvat/blob/zm/dm-cli-and-docs-update/datumaro/docs/user_manual.md#extract-subproject

ActiveChooN commented:
1. After import project from coco 

```(bash)
python -m datumaro project import -i ~/Downloads/test-ds -o ~dm-test-project -f coco --overwrite
```

I'm trying to extract subproject but get an error

```(bash)
python -m datumaro project extract -p ~/dm-test-project -o ~/dm-test-project-extractor -e '/item[image/width < image/height]'
2020-01-23 16:17:42,422 ERROR: local variable 'dst_dir' referenced before assignment
Traceback (most recent call last):
  File "/usr/lib/python3.5/runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/dkalinin/cvat/datumaro/datumaro/__main__.py", line 12, in <module>
    sys.exit(main())
  File "/home/dkalinin/cvat/datumaro/datumaro/cli/__main__.py", line 99, in main
    return args.command(args)
  File "/home/dkalinin/cvat/datumaro/datumaro/cli/contexts/project/__init__.py", line 366, in extract_command
    dst_dir = osp.abspath(dst_dir)
UnboundLocalError: local variable 'dst_dir' referenced before assignment
```
2. Propably it can be useful to add example transformer script

3. Get error when trying to add source folder with command arguments order like in `--help`

```(bash)
python -m datumaro source add -p ~/dm-test-project --name new-source -f image_dir path ~/Downloads/new-source 
usage: datumaro source add [-h] SOURCE_TYPE ...
datumaro source add: error: argument SOURCE_TYPE: invalid choice: '/home/dkalinin/dm-test-project' (choose from 'path', 'git')

```

zhiltsov-max commented:
@nmanovic, fixed the issue with VOC2007
@azhavoro, 1-2 fixed, 3. added
@ActiveChooN, 1. fixed, 2. transform is hidden while not implemented, 3. this invocation form is removed from the docs

nmanovic commented:
@zhiltsov-max , got the error again.
```
======================================================================
5651FAIL: test_can_save_and_load (tests.test_cvat_format.CvatConverterTest)
5652----------------------------------------------------------------------
5653Traceback (most recent call last):
5654  File "/home/django/datumaro/tests/test_cvat_format.py", line 268, in test_can_save_and_load
5655    target_dataset=DstTestExtractor())
5656  File "/home/django/datumaro/tests/test_cvat_format.py", line 178, in _test_save_and_load
5657    self.assertEqual(item_a, item_b, str(idx))
5658AssertionError: <datumaro.components.extractor.DatasetItem object at 0x7f9c7a6149b0> != <datumaro.components.project.DatasetItemWrapper object at 0x7f9c7a65dcf8> : 0
5659
5660
```

nmanovic commented on datumaro/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+
+![cli-design-image](docs/images/cli_design.png)
+
+- [FreeMind tool link](http://freemind.sourceforge.net/wiki/index.php/Main_Page)
```
Need to write something like: the diagram above was created by FreeMind tool.

nmanovic commented on datumaro/CONTRIBUTING.md:
````diff
@@ -0,0 +1,119 @@
+```
+├── [datumaro module]
+└── [project folder]
+    ├── .datumaro/
````
The project structure doesn't correspond to the current state. I'm OK if we fix the actual project structure later.
